controllers/release/release_adapter_test.go

```diff
@@ -0,0 +1,421 @@
+		result, err := adapter.EnsureFinalizersAreCalled()
+		Expect(result.CancelRequest).To(BeFalse())
+		Expect(err).NotTo(HaveOccurred())
+
+		// Add finalizer
+		result, err = adapter.EnsureFinalizerIsAdded()
+		Expect(result.CancelRequest).To(BeFalse())
+		Expect(err).NotTo(HaveOccurred())
+
+		// Create a ReleasePipelineRun to ensure it gets finalized
+		result, err = adapter.EnsureReleasePipelineRunExists()
+		Expect(result.CancelRequest).To(BeFalse())
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(k8sClient.Delete(ctx, release)).NotTo(HaveOccurred())
+		Eventually(func() bool {
+			err = k8sClient.Get(ctx, types.NamespacedName{
+				Name:      release.Name,
+				Namespace: release.Namespace,
+			}, release)
+			return err == nil && release.GetDeletionTimestamp() != nil
+		}, time.Second*10).Should(BeTrue())
+
+		result, err = adapter.EnsureFinalizersAreCalled()
+		Expect(result.RequeueRequest).To(BeTrue())
+		Expect(err).NotTo(HaveOccurred())
+
+		Eventually(func() error {
+			return k8sClient.Get(ctx, types.NamespacedName{
+				Name:      release.Name,
+				Namespace: release.Namespace,
+			}, release)
+		}, time.Second*10).Should(HaveOccurred())
+
+		Eventually(func() bool {
+			pipelineRun, _ := adapter.getReleasePipelineRun()
+			return (pipelineRun == nil)
+		}, time.Second*10).Should(BeTrue())
+	})
+
+	It("ensures Finalizers are called and fails for a Release object when the PipelineRun was already finalized", func() {
+		// Call to EnsureFinalizersAreCalled should succeed if the Release was not set to be deleted
+		result, err := adapter.EnsureFinalizersAreCalled()
+		Expect(result.CancelRequest).To(BeFalse())
+		Expect(err).NotTo(HaveOccurred())
+
+		// Create a ReleasePipelineRun to ensure it gets finalized
+		result, err = adapter.EnsureReleasePipelineRunExists()
+		Expect(result.CancelRequest).To(BeFalse())
+		Expect(err).NotTo(HaveOccurred())
+
+		// Add finalizer
+		result, err = adapter.EnsureFinalizerIsAdded()
+		Expect(result.CancelRequest).To(BeFalse())
+		Expect(err).NotTo(HaveOccurred())
+
+		result, err = adapter.EnsureReleasePipelineStatusIsTracked()
+		Expect(result.CancelRequest).To(BeFalse())
+		Expect(err).NotTo(HaveOccurred())
+
+		release.SetDeletionTimestamp(&metav1.Time{Time: time.Now()})
+
+		// delete the pipelineRun before the release so it should throw an
+		// error on finalizing the release
+		pipelineRun, _ := adapter.getReleasePipelineRun()
+		Expect(k8sClient.Delete(ctx, pipelineRun)).Should(Succeed())
+		Expect(k8sClient.Delete(ctx, release)).Should(Succeed())
+
+		result, err = adapter.EnsureFinalizersAreCalled()
+		Expect(!result.CancelRequest && err != nil).To(BeTrue())
+	})
+
+	It("can create a ReleasePipelineRun", func() {
+		pipelineRun, err := adapter.createReleasePipelineRun(
+			releaseStrategy,
+			applicationSnapshot)
+		Expect(err).ShouldNot(HaveOccurred())
+		Expect(pipelineRun.Spec.ServiceAccountName).Should(ContainSubstring("test-account"))
```

Comment:
Missing checks:

- Owner.
- Release labels.
- Params for the snapshot and for params in the ReleaseStrategy.
- Workspace.

Reply:
Just wondering here if I do have to test all of this here or should I rewrite the test for `tekton.NewReleasePipelineRun` testing these things. it something goes wrong with them it should be something wrong on the tekton package, not the adapter (in this case, I guess).

It seems more appropriate as here my Goal is only test the Status. Are these attributes part of the Status as well?

Reply:
Here. This test is about creating a ReleasePipelineRun, and a ReleasePipelineRun is basically a PipelineRun, with an owner, release labels and all the other stuff.

Reply:
true. I was looking into another one. I guess due to me moving tests around I thought your comments where about the  `can add the release information to its status and mark it as Running`

Reply:
I will add these changes, although I still think this should be assured in the `tekton.NewReleasePipeline()` test, not here. BTW, it is not done there and should be fixed (I will do when I finish here).

Reply:
Nope. It shouldn't be done there. NewReleasePipeline is an entry point to create a ReleasePipelineRun. It creates the bare minimum. Here, we compose the ReleasePipelineRun adding everything that is needed. `createReleasePipelineRun` should contain calls to all those functions to generate a valid ReleasePipelineRun. If in the future any of those calls are removed by mistake, we want to know we screwed up, because that's a big bug.